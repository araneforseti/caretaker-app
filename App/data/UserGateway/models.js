const Permissions = {
    FOCUS: 'focus',
    ADMIN: 'admin',
    HELPER: 'helper'
}

const NotificationTypes = {
    PUSH: 'push',
    EMAIL: 'email',
    SMS: 'sms'
}

const NotificationTriggers = {
    ANNOUNCEMENT: 'announcement',
    EVENT: 'event',
    EVENT_BY_ROLE: 'event by role'
}

const Settings = {
    [NotificationTypes.PUSH]: {
        name: 'Push',
        value: true
    },
    [NotificationTypes.EMAIL]: {
        name: 'Email',
        value: false
    },
    [NotificationTypes.SMS]: {
        name: 'SMS',
        value: false
    },
    [NotificationTriggers.ANNOUNCEMENT]: {
        name: 'Announcements',
        value: true
    },
    [NotificationTriggers.EVENT]: {
        name: 'Events',
        value: false
    },
    [NotificationTriggers.EVENT_BY_ROLE]: {
        name: 'Events specific to your role',
        value: true
    }
}

module.exports = {
  Permissions: Permissions,
  NotificationTypes: NotificationTypes,
  NotificationTriggers: NotificationTriggers,
  Settings: Settings
}
