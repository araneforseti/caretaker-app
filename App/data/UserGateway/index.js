import { Record } from 'immutable';
import UserModels from './models.js'
import { API_URL } from '../../config/ApiConfig';

export const Permissions = UserModels.Permissions
export const NotificationTypes = UserModels.NotificationTypes
export const NotificationTriggers = UserModels.NotificationTriggers

const createUserSettings = () => ({
  [NotificationTypes.PUSH]: {
    name: 'Push',
    value: true
  },
  [NotificationTypes.EMAIL]: {
    name: 'Email',
    value: false
  },
  [NotificationTypes.SMS]: {
    name: 'SMS',
    value: false
  },
  [NotificationTriggers.ANNOUNCEMENT]: {
    name: 'Announcements',
    value: true
  },
  [NotificationTriggers.EVENT]: {
    name: 'Events',
    value: false
  },
  [NotificationTriggers.EVENT_BY_ROLE]: {
    name: 'Events specific to your role',
    value: true
  }
})

export const User = Record({
  name: 'Defaul User',
  identifier: 'default@domain.com',
  permission: Permissions.HELPER,
  roles: [],
  phone: '123-456-7890',
  settings: createUserSettings()
});

const createUserGateway = function() {
  let allUsers = null;
  let currentUser = null;

  return {
    getCurrentUser() {
      return currentUser;
    },

    getAll() {
      if (allUsers === null) {
        //api call goes here
        allUsers = [
          new User({ name: 'S', identifier: 's', permission: Permissions.FOCUS, roles: [], phone: '123-456-7890', settings: createUserSettings() }),
          new User({ name: 'Sarah', identifier: 'sarah@emailprovider.com', permission: Permissions.FOCUS, roles: [], phone: '123-456-7890', settings: createUserSettings() }),
          new User({ name: 'Caroline', identifier: 'caroline@woahdude.com', permission: Permissions.ADMIN, roles: ['driver', 'coordinator'], phone: '123-456-7890', settings: createUserSettings() }),
          new User({ name: 'Jack', identifier: 'jack@coolwebsite.com', permission: Permissions.HELPER, roles: ['active friend'], settings: createUserSettings() }),
          new User({ name: 'Austin', identifier: 'austin@yeehaw.com', permission: Permissions.HELPER, roles: [], phone: '123-456-7890', settings: createUserSettings() })
        ]
      }
      return allUsers;
    },

    getName(identifier) {
      const user = allUsers.find((user) => user.identifier === identifier)
      return user ? user.name : 'TBD';
    },

    getRoles() {
      return currentUser.roles
    },

    login(email) {
      //make call to google oauth and store the results
      let results = {
        email: email
      };
      let userFromBackend = this.getAll().find((user) => user.identifier === results.email);
      if (typeof userFromBackend !== 'undefined') {
        currentUser = userFromBackend;
        return true;
      }
      return false;
    },

    logout() {
      // make call to oauth and/or invalidate current session
      this.currentUser = null;
    },

    isSelf(identifier) {
      return currentUser.identifier === identifier;
    },

    isAdmin() {
      return currentUser.permission === Permissions.ADMIN ||
        currentUser.permission === Permissions.FOCUS;
    },

    getCurrentPermissions() {
      return currentUser.permission;
    },

    inviteUser(email) {
      //api call to invite the user
      alert('invited ' + email + '!');
    },

    updateSettings(settings) {
      //api call to update the user settings
    }
  }
}

export default createUserGateway();
