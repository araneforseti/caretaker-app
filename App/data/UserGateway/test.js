import UserGateway from './index.js';
import { Record } from 'immutable';

describe('UserGateway', () => {

    it('should return false when attempting to sign in with a non-existent user', () => {
        expect(UserGateway.login('TBD')).toBe(false);
    });

    it('should return true when sign in is successful', () => {
        expect(UserGateway.login('sarah@emailprovider.com')).toBe(true);
    });

    it('should assign TBD as a name when TBD is the identifier', () => {
        expect(UserGateway.getName('TBD')).toBe('TBD');
    });

    it('should get the current user as an immutable record', () => {
        UserGateway.login('s');

        const user =  UserGateway.getCurrentUser();

        //Record.isRecord is in the immutable@4.0.0 release candidate
        expect(user instanceof Record).toBe(true);
    })
});
