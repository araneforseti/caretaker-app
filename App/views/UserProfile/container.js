import React, { Component } from 'react';
import UserGateway, { NotificationTypes } from '../../data/UserGateway/';
import ScreenWithToolbar from '../../components/shared/ScreenWithToolbar';
import { Settings, User } from '../../data/UserGateway';

export default ContainerFor = (UserProfile) => class extends Component {
    constructor() {
        super();
        const defaultUser = new User();
        this.state = {
            currentUser: defaultUser
        }
    }

    componentDidMount() {
        const user = UserGateway.getCurrentUser()
        this.setState({
            currentUser: user
        });
    }

    toggle = (key) => () => {
        let settings = this.state.currentUser.settings;
        const currentValue = settings[key].value
        settings[key].value = !currentValue;
        UserGateway.updateSettings(settings);
        this.setState({
            currentUser: this.state.currentUser.set('settings', settings)
        });
    }

    render() {
        const user = this.state.currentUser;
        return <ScreenWithToolbar title='Profile' navigation={this.props.navigation}>
            <UserProfile
                name={user.name}
                email={user.identifier}
                phone={user.phone}
                settings={user.settings}
                toggle={this.toggle}
            />
        </ScreenWithToolbar>
    }
}
