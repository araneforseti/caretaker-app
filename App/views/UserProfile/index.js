import React, { Component } from 'react';
import {
  StyleSheet,
  Text,
  View
} from 'react-native';
import ContainerFor from './container.js';
import Checkbox from '../../components/material-ui/Checkbox.js';
import { NotificationTypes, NotificationTriggers } from '../../data/UserGateway';

const UserProfile = (props) => {

    _renderCheckbox = (key) => {
        const currentValue = props.settings[key].value;

        return <Checkbox
            label={props.settings[key].name}
            value={key}
            checked={currentValue}
            onCheck={props.toggle(key)}
        />
    }

    return <View style={styles.container}>
        <Text style={styles.title}>User Profile</Text>
        <View style={styles.sectionContainer}>
            <Text style={styles.sectionTitle}>Info</Text>
            <Text>Name: {props.name}</Text>
            <Text>Email: {props.email}</Text>
            <Text>Phone: {props.phone}</Text>
        </View>
        <View style={styles.sectionContainer}>
            <Text style={styles.sectionTitle}>Receive Notifications via:</Text>
            {_renderCheckbox(NotificationTypes.PUSH)}
            {_renderCheckbox(NotificationTypes.EMAIL)}
            {_renderCheckbox(NotificationTypes.SMS)}
        </View>
        <View style={styles.sectionContainer}>
            <Text style={styles.sectionTitle}>Receive Notifications for:</Text>
            {_renderCheckbox(NotificationTriggers.ANNOUNCEMENT)}
            {_renderCheckbox(NotificationTriggers.EVENT)}
            {_renderCheckbox(NotificationTriggers.EVENT_BY_ROLE)}
        </View>
    </View>
};

export default ContainerFor(UserProfile);

const styles = StyleSheet.create({
    button: {
        margin: 10,
        padding: 10
    },
    container: {
        flex: 1,
        backgroundColor: '#F5FCFF',
        justifyContent: 'flex-start'
    },
    title: {
        fontSize: 24,
        textAlign: 'center',
        margin: 15,
    },
    sectionContainer: {
        margin: 10,
        flex: 1
    },
    sectionTitle: {
        fontSize: 16,
        fontWeight: 'bold',
        marginTop: 10,
    }
});
