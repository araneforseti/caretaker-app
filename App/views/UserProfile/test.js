import React, { Component } from 'react';
import renderer from 'react-test-renderer';
import UserGateway, { NotificationTypes, NotificationTriggers } from '../../data/UserGateway/';
import UserProfile from './index.js';
import { Record } from 'immutable';

describe('User Profile', () => {

    beforeEach(() => {
        UserGateway.login('s');
    });

    it('snapshot', () => {
        expect(renderer.create(<UserProfile />)).toMatchSnapshot();
    })

    it('should get the user settings from the gateway', () => {
        const profile = renderer.create(<UserProfile />).getInstance();
        expect(profile.state.currentUser.settings).toEqual({
            [NotificationTypes.PUSH]: {
                name: 'Push',
                value: true
            },
            [NotificationTypes.EMAIL]: {
                name: 'Email',
                value: false
            },
            [NotificationTypes.SMS]: {
                name: 'SMS',
                value: false
            },
            [NotificationTriggers.ANNOUNCEMENT]: {
                name: 'Announcements',
                value: true
            },
            [NotificationTriggers.EVENT]: {
                name: 'Events',
                value: false
            },
            [NotificationTriggers.EVENT_BY_ROLE]: {
                name: 'Events specific to your role',
                value: true
            }
        });
    })

    it('should toggle the notification types', () => {
        const profile = renderer.create(<UserProfile />).getInstance();

        profile.toggle(NotificationTypes.EMAIL)();

        expect(profile.state.currentUser.settings).toEqual({
            [NotificationTypes.PUSH]: {
                name: 'Push',
                value: true
            },
            [NotificationTypes.EMAIL]: {
                name: 'Email',
                value: true
            },
            [NotificationTypes.SMS]: {
                name: 'SMS',
                value: false
            },
            [NotificationTriggers.ANNOUNCEMENT]: {
                name: 'Announcements',
                value: true
            },
            [NotificationTriggers.EVENT]: {
                name: 'Events',
                value: false
            },
            [NotificationTriggers.EVENT_BY_ROLE]: {
                name: 'Events specific to your role',
                value: true
            }
        });
    })

    it('should update the user settings through the gateway', () => {
        const spy = jest.spyOn(UserGateway, 'updateSettings');
        const profile = renderer.create(<UserProfile />).getInstance();

        profile.toggle(NotificationTypes.EMAIL)();

        expect(spy).toHaveBeenCalledWith(profile.state.currentUser.settings);
    })
})
