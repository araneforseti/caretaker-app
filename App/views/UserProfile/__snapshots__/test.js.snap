// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Profile snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <ThemeProvider
    uiTheme={
      Object {
        "palette": Object {
          "accentColor": "#FF7575",
          "alternateTextColor": "white",
          "lightColor": "#FFC0C0",
          "primaryColor": "#FF7575",
          "primaryTextColor": "black",
          "secondaryColor": "#CC5D5D",
        },
        "spacing": Object {
          "miniActionButtonSize": 42,
        },
        "toolbar": Object {
          "container": Object {
            "height": 50,
          },
        },
        "typography": Object {
          "fontSize": Object {
            "display": 25,
          },
          "subheading": Object {
            "fontSize": 16,
            "fontWeight": "normal",
            "lineHeight": 24,
          },
        },
      }
    }
  />
  <View
    style={
      Object {
        "backgroundColor": "#F5FCFF",
        "flex": 1,
        "justifyContent": "flex-start",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 24,
          "margin": 15,
          "textAlign": "center",
        }
      }
    >
      User Profile
    </Text>
    <View
      style={
        Object {
          "flex": 1,
          "margin": 10,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 16,
            "fontWeight": "bold",
            "marginTop": 10,
          }
        }
      >
        Info
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        Name: 
        S
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        Email: 
        s
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        Phone: 
        123-456-7890
      </Text>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "margin": 10,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 16,
            "fontWeight": "bold",
            "marginTop": 10,
          }
        }
      >
        Receive Notifications via:
      </Text>
      <ThemeProvider
        uiTheme={
          Object {
            "palette": Object {
              "accentColor": "#FF7575",
              "alternateTextColor": "white",
              "lightColor": "#FFC0C0",
              "primaryColor": "#FF7575",
              "primaryTextColor": "black",
              "secondaryColor": "#CC5D5D",
            },
            "spacing": Object {
              "miniActionButtonSize": 42,
            },
            "toolbar": Object {
              "container": Object {
                "height": 50,
              },
            },
            "typography": Object {
              "fontSize": Object {
                "display": 25,
              },
              "subheading": Object {
                "fontSize": 16,
                "fontWeight": "normal",
                "lineHeight": 24,
              },
            },
          }
        }
      />
      <ThemeProvider
        uiTheme={
          Object {
            "palette": Object {
              "accentColor": "#FF7575",
              "alternateTextColor": "white",
              "lightColor": "#FFC0C0",
              "primaryColor": "#FF7575",
              "primaryTextColor": "black",
              "secondaryColor": "#CC5D5D",
            },
            "spacing": Object {
              "miniActionButtonSize": 42,
            },
            "toolbar": Object {
              "container": Object {
                "height": 50,
              },
            },
            "typography": Object {
              "fontSize": Object {
                "display": 25,
              },
              "subheading": Object {
                "fontSize": 16,
                "fontWeight": "normal",
                "lineHeight": 24,
              },
            },
          }
        }
      />
      <ThemeProvider
        uiTheme={
          Object {
            "palette": Object {
              "accentColor": "#FF7575",
              "alternateTextColor": "white",
              "lightColor": "#FFC0C0",
              "primaryColor": "#FF7575",
              "primaryTextColor": "black",
              "secondaryColor": "#CC5D5D",
            },
            "spacing": Object {
              "miniActionButtonSize": 42,
            },
            "toolbar": Object {
              "container": Object {
                "height": 50,
              },
            },
            "typography": Object {
              "fontSize": Object {
                "display": 25,
              },
              "subheading": Object {
                "fontSize": 16,
                "fontWeight": "normal",
                "lineHeight": 24,
              },
            },
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "margin": 10,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 16,
            "fontWeight": "bold",
            "marginTop": 10,
          }
        }
      >
        Receive Notifications for:
      </Text>
      <ThemeProvider
        uiTheme={
          Object {
            "palette": Object {
              "accentColor": "#FF7575",
              "alternateTextColor": "white",
              "lightColor": "#FFC0C0",
              "primaryColor": "#FF7575",
              "primaryTextColor": "black",
              "secondaryColor": "#CC5D5D",
            },
            "spacing": Object {
              "miniActionButtonSize": 42,
            },
            "toolbar": Object {
              "container": Object {
                "height": 50,
              },
            },
            "typography": Object {
              "fontSize": Object {
                "display": 25,
              },
              "subheading": Object {
                "fontSize": 16,
                "fontWeight": "normal",
                "lineHeight": 24,
              },
            },
          }
        }
      />
      <ThemeProvider
        uiTheme={
          Object {
            "palette": Object {
              "accentColor": "#FF7575",
              "alternateTextColor": "white",
              "lightColor": "#FFC0C0",
              "primaryColor": "#FF7575",
              "primaryTextColor": "black",
              "secondaryColor": "#CC5D5D",
            },
            "spacing": Object {
              "miniActionButtonSize": 42,
            },
            "toolbar": Object {
              "container": Object {
                "height": 50,
              },
            },
            "typography": Object {
              "fontSize": Object {
                "display": 25,
              },
              "subheading": Object {
                "fontSize": 16,
                "fontWeight": "normal",
                "lineHeight": 24,
              },
            },
          }
        }
      />
      <ThemeProvider
        uiTheme={
          Object {
            "palette": Object {
              "accentColor": "#FF7575",
              "alternateTextColor": "white",
              "lightColor": "#FFC0C0",
              "primaryColor": "#FF7575",
              "primaryTextColor": "black",
              "secondaryColor": "#CC5D5D",
            },
            "spacing": Object {
              "miniActionButtonSize": 42,
            },
            "toolbar": Object {
              "container": Object {
                "height": 50,
              },
            },
            "typography": Object {
              "fontSize": Object {
                "display": 25,
              },
              "subheading": Object {
                "fontSize": 16,
                "fontWeight": "normal",
                "lineHeight": 24,
              },
            },
          }
        }
      />
    </View>
  </View>
</View>
`;
